File: Partner-Api/src/main/java/work/slhaf/partner/api/factory/entity/AgentRegisterContext.java
package work.slhaf.partner.api.factory.entity;

import lombok.Data;
import org.reflections.Reflections;
import org.reflections.scanners.Scanners;
import org.reflections.util.ConfigurationBuilder;
import work.slhaf.partner.api.common.chat.pojo.Message;
import work.slhaf.partner.api.factory.config.pojo.ModelConfig;
import work.slhaf.partner.api.factory.module.pojo.MetaModule;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

@Data
public class AgentRegisterContext {
    private Reflections reflections;
    private final HashMap<String, Function<Object[], Object>> methodsRouterTable = new HashMap<>();
    private final HashMap<String, Function<Object[], Object>> coordinatedMethodsRouterTable = new HashMap<>();
    private final HashMap<Class<?>, Object> capabilityCoreInstances = new HashMap<>();
    private final HashMap<Class<?>, Object> capabilityHolderInstances = new HashMap<>();
    private Set<Class<?>> cores;
    private Set<Class<?>> capabilities;
    private HashMap<String, List<Message>> modelPromptMap = new HashMap<>();
    private HashMap<String, ModelConfig> modelConfigMap = new HashMap<>();
    private List<MetaModule> moduleList = new ArrayList<>();

    public AgentRegisterContext(List<String> paths) {
        reflections = new Reflections(new ConfigurationBuilder().setScanners(
                        Scanners.FieldsAnnotated,
                        Scanners.SubTypes,
                        Scanners.MethodsAnnotated,
                        Scanners.TypesAnnotated
                )
                .forPackages(paths.toArray(paths.toArray(new String[0])))
        );
    }
}

File: Partner-Framework/src/main/java/work/slhaf/partner/framework/agent/factory/context/AgentRegisterContext.kt
package work.slhaf.partner.framework.agent.factory.context

import org.reflections.Reflections
import org.reflections.scanners.Scanners
import org.reflections.util.ConfigurationBuilder
import java.lang.reflect.Method
import java.net.URL

class AgentRegisterContext(urls: List<URL>) {
    val reflections: Reflections = Reflections(
        ConfigurationBuilder().setScanners(
            Scanners.FieldsAnnotated,
            Scanners.SubTypes,
            Scanners.MethodsAnnotated,
            Scanners.TypesAnnotated
        ).setUrls(urls)
    )

    val capabilityFactoryContext: CapabilityFactoryContext = CapabilityFactoryContext()
    val componentFactoryContext: ComponentFactoryContext = ComponentFactoryContext()
    val agentContext: AgentContext = AgentContext
}

class CapabilityFactoryContext {
    val cores: MutableSet<Class<*>> = LinkedHashSet()
    val capabilities: MutableSet<Class<*>> = LinkedHashSet()
    val methods: MutableSet<Method> = LinkedHashSet()
}

class ComponentFactoryContext {
    val initMethodsByDeclaringType: MutableMap<Class<*>, MutableSet<Method>> = LinkedHashMap()
}
